docs/design-documents/20220408-loading-pipeline-configuration-files.md

```diff
@@ -0,0 +1,282 @@
+# Loading Pipeline Configuration Files
+
+## Summary 
+
+This document describes the design and implementation for loading pipeline 
+configuration files. It details how Conduit loads configuration files at start.
+
+In order to provide a detailed history of configuration changes, better secret
+handling, and increased confidence in a given configuration, we propose
+adding support for loading and subsequently running pipelines entirely from 
+configuration files loaded into Conduit from a configured directory.
+
+Our solution proposes an immutable Conduit runtime backed by an in-memory DB
```

Comment:
I'm not sure I follow the reasoning behind the in-memory DB. IIUC we would like to build the pipeline from the config file and keep it only in memory without persisting it anywhere, right? At the same time the document mentions that the `ConnectorPersister` would still be able to persist the connector state (position) between restarts. If that's what we are trying to do, it's not going to work that easily, because the connector persister is persisting the whole connector and not only its state. We would need to make sure that we can (a) persist the connector state separately and (b) that we can associate the state with a specific connector.

To solve problem (a) we could persist the pipeline, connectors and processors same as if they were created through the API and update them when the pipeline config is loaded. Then the config file only plays a role in how the pipeline is created, everything else functions the same.

To solve problem (b) we need a way to associate the persisted pipeline with a certain pipeline config file. If we persist pipelines in the store (as proposed in the previous paragraph) then the simple solution is to add a field to the pipeline config file that uniquely identifies a single pipeline (i.e. user explicitly sets the pipeline id). Otherwise we would need to allow the user to uniquely identify every single connector so that we can associate it to the stored state.